qiskit/transpiler/layout.py

```diff
@@ -374,3 +374,4 @@ class TranspileLayout:
 
     initial_layout: Layout
     input_qubit_mapping: Dict[Qubit, int]
+    final_layout: Optional[Layout] = None
```

Comment:
What does it mean for `final_layout` to be optional? Is it that transpiler passes could swap qubits and not report it, or that they can leave this unset if they are not performing any swaps?

Reply:
It primarily means the latter, if `final_layout` is None then there wasn't a permutation being caused by the routing stage in the compiler. Basically, the routing pass was skipped. There is also a backwards compatibility component to making this optional as well, but that's less likely to come up in practice outside of the routing pass not run.